### ISSUE TITLE
CUDA OOM when `rescore_with_attention_decoder`

### ISSUE THREAD

**Issue by wgb14** [k2-fsa/icefall#263]
See log below.
```
2022-03-21 12:55:28,824 INFO [decode.py:735] num_arcs before pruning: 448886
2022-03-21 12:55:28,824 INFO [decode.py:738] This OOM is not an error. You can ignore it. If your model does not converge well, or --max-duration is too large, or the input sound file is difficult to decode, you will meet this exception.
2022-03-21 12:55:28,879 INFO [decode.py:748] num_arcs after pruning: 13687
Traceback (most recent call last):
  File "./conformer_ctc/decode.py", line 705, in <module>
    main()
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "./conformer_ctc/decode.py", line 691, in main
    eos_id=eos_id,
  File "./conformer_ctc/decode.py", line 452, in decode_dataset
    eos_id=eos_id,
  File "./conformer_ctc/decode.py", line 371, in decode_one_batch
    nbest_scale=params.nbest_scale,
  File "/userhome/user/guanbo/icefall/icefall/decode.py", line 828, in rescore_with_attention_decoder
    nbest = nbest.intersect(lattice)
  File "/userhome/user/guanbo/icefall/icefall/decode.py", line 330, in intersect
    sorted_match_a=True,
  File "/userhome/user/guanbo/icefall/icefall/decode.py", line 64, in _intersect_device
    return k2.cat(ans)
  File "/opt/conda/lib/python3.7/site-packages/k2-1.14.dev20220321+cuda11.1.torch1.10.0-py3.7-linux-x86_64.egg/k2/ops.py", line 218, in cat
    ans_ragged_arcs = _k2.cat(src_ragged_arcs, axis=0)
RuntimeError: CUDA out of memory. Tried to allocate 6.57 GiB (GPU 0; 31.75 GiB total capacity; 25.97 GiB already allocated; 481.75 MiB free; 29.92 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```
I think this is a similar situation to `rescore_with_whole_lattice` https://github.com/k2-fsa/icefall/blob/b2b4d9e0b60b65d6ecd40cc252d60930784bb830/icefall/decode.py#L719-L759
and we can avoid this error being raised.
I can think of 2 options when CUDA OOM detected:
 - prune the rescored lattice
 - decrease `num_paths`

Any insights?

**Comment by wgb14** [comment]
I can think of 2 options when CUDA OOM detected:
 - prune the rescored lattice
 - decrease `num_paths`

Any insights?

**Comment by csukuangfj** [comment]
The easiest way may be to decrease `num_paths` while creating the nbest object.

**Comment by wgb14** [comment]
Thanks! decreasing `num_paths` works well.

**Comment by wgb14** [comment]
Solved in #120. Closing this
